# 알바이오 첫번째 발표

- 발표자료를 참고하시면 도움이 됩니다

### SeqinR 패키지 설치하고 불러와보기

In [3]:
#install.packages("seqinr")
library(seqinr)

# SeqinR 패키지 불러오기

In [2]:
library(seqinr)

In [3]:
lseqinr()

[1] "a"                       "aaa"                    
  [3] "AAstat"                  "acnucclose"             
  [5] "acnucopen"               "al2bp"                  
  [7] "alllistranks"            "alr"                    
  [9] "amb"                     "as.alignment"           
 [11] "as.matrix.alignment"     "as.SeqAcnucWeb"         
 [13] "as.SeqFastaAA"           "as.SeqFastadna"         
 [15] "as.SeqFrag"              "autosocket"             
 [17] "baselineabif"            "bma"                    
 [19] "c2s"                     "cai"                    
 [21] "cfl"                     "choosebank"             
 [23] "circle"                  "clfcd"                  
 [25] "clientid"                "closebank"              
 [27] "col2alpha"               "comp"                   
 [29] "computePI"               "con"                    
 [31] "consensus"               "count"                  
 [33] "countfreelists"          "countsubseqs"           
 [35] "crelistfromclientdata"   "css"                    
 [37] "dia.bactgensize"         "dia.db.growth"          
 [39] "dist.alignment"          "dotchart.uco"           
 [41] "dotPlot"                 "draw.oriloc"            
 [43] "draw.rearranged.oriloc"  "draw.recstat"           
 [45] "exseq"                   "extract.breakpoints"    
 [47] "extractseqs"             "fastacc"                
 [49] "gb2fasta"                "gbk2g2"                 
 [51] "gbk2g2.euk"              "GC"                     
 [53] "GC1"                     "GC2"                    
 [55] "GC3"                     "GCpos"                  
 [57] "getAnnot"                "getAnnot.default"       
 [59] "getAnnot.list"           "getAnnot.logical"       
 [61] "getAnnot.qaw"            "getAnnot.SeqAcnucWeb"   
 [63] "getAnnot.SeqFastaAA"     "getAnnot.SeqFastadna"   
 [65] "getAttributsocket"       "get.db.growth"          
 [67] "getFrag"                 "getFrag.character"      
 [69] "getFrag.default"         "getFrag.list"           
 [71] "getFrag.logical"         "getFrag.qaw"            
 [73] "getFrag.SeqAcnucWeb"     "getFrag.SeqFastaAA"     
 [75] "getFrag.SeqFastadna"     "getFrag.SeqFrag"        
 [77] "getKeyword"              "getKeyword.default"     
 [79] "getKeyword.list"         "getKeyword.logical"     
 [81] "getKeyword.qaw"          "getKeyword.SeqAcnucWeb" 
 [83] "getLength"               "getLength.character"    
 [85] "getLength.default"       "getLength.list"         
 [87] "getLength.logical"       "getLength.qaw"          
 [89] "getLength.SeqAcnucWeb"   "getLength.SeqFastaAA"   
 [91] "getLength.SeqFastadna"   "getLength.SeqFrag"      
 [93] "getlistrank"             "getliststate"           
 [95] "getLocation"             "getLocation.default"    
 [97] "getLocation.list"        "getLocation.logical"    
 [99] "getLocation.qaw"         "getLocation.SeqAcnucWeb"
[101] "getName"                 "getName.default"        
[103] "getName.list"            "getName.logical"        
[105] "getName.qaw"             "getName.SeqAcnucWeb"    
[107] "getName.SeqFastaAA"      "getName.SeqFastadna"    
[109] "getName.SeqFrag"         "get.ncbi"               
[111] "getNumber.socket"        "getSequence"            
[113] "getSequence.character"   "getSequence.default"    
[115] "getSequence.list"        "getSequence.logical"    
[117] "getSequence.qaw"         "getSequence.SeqAcnucWeb"
[119] "getSequence.SeqFastaAA"  "getSequence.SeqFastadna"
[121] "getSequence.SeqFrag"     "getTrans"               
[123] "getTrans.character"      "getTrans.default"       
[125] "getTrans.list"           "getTrans.logical"       
[127] "getTrans.qaw"            "getTrans.SeqAcnucWeb"   
[129] "getTrans.SeqFastadna"    "getTrans.SeqFrag"       
[131] "getType"                 "gfrag"                  
[133] "ghelp"                   "gln"                    
[135] "glr"                     "gls"                    
[137] "isenum"                  "isn"                    


### Bioconductor 설치 및 실행

In [1]:
source("http://bioconductor.org/biocLite.R")
biocLite()

Installing package into ‘/home/deepbio/R/i686-pc-linux-gnu-library/3.2’
(as ‘lib’ is unspecified)



The downloaded source packages are in
	‘/tmp/RtmpLqiTUY/downloaded_packages’


Bioconductor version 3.2 (BiocInstaller 1.20.3), ?biocLite for help
A new version of Bioconductor is available after installing the most recent
  version of R; see http://bioconductor.org/install
BioC_mirror: https://bioconductor.org
Using Bioconductor 3.2 (BiocInstaller 1.20.3), R 3.2.3 (2015-12-10).
Installing package(s) ‘Biobase’, ‘IRanges’, ‘AnnotationDbi’
also installing the dependencies ‘BiocGenerics’, ‘S4Vectors’, ‘DBI’, ‘RSQLite’




The downloaded source packages are in
	‘/tmp/RtmpLqiTUY/downloaded_packages’


Old packages: 'boot', 'cluster', 'Matrix', 'mgcv', 'nlme', 'survival'


### SeqinR 을 이용하여 NCBI 지놈 데이터를 가져오는 R 스크립트

- 스크립트가 돌아가질 않습니다, 아마도 튜토리얼 시기에 만들어놓은 라이브러리가 지금의 라이브러리와는 살짝 달라서 그렇지 않나 싶습니다

- 그래서 일단 FASTA 파일을 Write 한 뒤에 읽어들여 보겠습니다

In [8]:
getncbiseq <- function(accession) # Fucntion 을 만들고
  {
    library(seqinr) # 이 함수는 seqinr 패키지를 필요로 하기 때문에 require 문을 사용하여 임포트함 
    # first find which ACNUC database the accession is stored in:
    dbs <- c("genbank","refseq","refseqViruses","bacterial")
    numdbs <- length(dbs)
    for (i in 1:numdbs)
    {
        db <- dbs[i]
        choosebank(db)
        # check if the sequence is in ACNUC database 'db':
        resquery <- try(query(".tmpquery", paste("AC=", accession)), silent = TRUE)
        if (!(inherits(resquery, "try-error")))
        {
           queryname <- "query2"
           thequery <- paste("AC=",accession,sep="")
           query(`queryname`,`thequery`)
           # see if a sequence was retrieved:
           seq <- getSequence(query2$req[[1]])
           closebank()
           return(seq)
        }
        closebank()
     }
     print(paste("ERROR: accession",accession,"was not found")) # 만약에 시도한 값이 발견되지 않으면 에러를 리턴
  }

In [9]:
dengueseq <- getncbiseq("NC_001477")

원래 **denguseq[1:50]** 을 입력했을 시에 결과값이

![IMG](DNASEQ_RESULT.JPG)

이 나와야 정상입니다만, 나오질 않아서 다른 방법으로 시도해 보도록 하겠습니다

일단 FASTA 파일을 쓰고, 읽어보도록 하겠습니다

일단 파일을 읽는 함수 read.fasta(file = 파일이름)[[1]] 을 사용해보도록 하겠습니다

![IMG](how to use count.JPG)

이들이 이야기하는대로 다시 시도해 봅시다

여기에서 system 문은 사용하지 마시고 믿에 구문을 사용하면 됩니다

In [4]:
myseq <- read.fasta(file = "den1.fasta")[[1]]

In [10]:
myseq[1:50]

[1] "a" "g" "t" "t" "g" "t" "t" "a" "g" "t" "c" "t" "a" "c" "g" "t" "g" "g" "a"
[20] "c" "c" "g" "a" "c" "a" "a" "g" "a" "a" "c" "a" "g" "t" "t" "t" "c" "g" "a"
[39] "a" "t" "c" "g" "g" "a" "a" "g" "c" "t" "t" "g"

정상적으로 denguseq[1:50] 을 입력했을때의 결과값과 동일하게 출력되는것을 보실 수 있습니다



## DNA Sequnence 의 길이 구하기

여기에서는 length() 함수를 이용하면 됩니다, 정상적인 경우에는 **10735** 개라고 출력됩니다

In [10]:
length(myseq)

[1] 10735

### DNA Sequence 를 압축(Base Composition)하기

4개의 다른 뉴클레오타이드 형태("A", "C", "G", "T")로 압축하려면, table() 함수를 사용하면 쉽게 압축 할 수 있습니다

정상적인 경우에는 아래와 같이 출력되게 됩니다

In [11]:
table(myseq)

myseq
   a    c    g    t 
3426 2240 2770 2299 

### DNA 의 GC Content

이분들이 이야기하기를, GC Content는 Genome Sequncing 에서 가장 핵심적인 것이라고 할 수 있다고 합니다

### GC Content 란 무었일까?

우선 아데닌(A), 타이민(T), 구아닌(G), 사이토신(C) - 4 종류의 염기가 이 DNA 를 구성하는데, 이중 구아닌과 사이토신이 전체 DNA 에서 함유된 량을 GC% 라고 한다.

아데닌 타이민은 2개의 수소결합으로 이루어져 있지만, 구아닌과 사이토신은 3의 수소결합으로 이루어져 있어서 더 안정적이다

GC% 가 높을수록 DNA 는 안정적이지만, 수소결합 자체가 강한 결합은 아니므로 안정적인 결합이 아니라고 할 수 있다

다만 GC% 가 높을수록 DNA의 녹는 속도는 높아진다, DNA중 단백질을 발현시킬수 있는 코딩 부분은 다른 DNA 부분보다 GC% 가 높다

출처 네이버 지식백과 ~~네이버가 이렇게 좋은 정보를 준 적은 없어~~

### 계산 공식은?

공식은 일반적으로 G와 C를 더한 값 * 100을 한 후에 ACGT 를 더한 값과 나누어 줍니다

### 여기에서 GC 를 사용하려면? 

구지 일일히 계산할 필요 없이 gc() 함수를 사용하면 됩니다, 아래와 같이 입력하면 됩니다


In [12]:
GC(myseq)

[1] 0.4666977

그러면 여기에 100을 곱하면, 결과적으로 46.669771% 가 나오게 됩니다

여기에서 왜 100을 곱하는지 궁금해서 알아보았더니, Gs 혹은 Cs 값이 0.4666977 이 나오기 때문에 

Fraction 을 Percentage 로 변환해 주면, 46.669771% 이 나오게 되는 것입니다

따라서 Percentage of GC Content 는 46.669771% 이라는 것을 알수 있게 됩니다

### DNA Words 

DNA 에서 기본적으로 A(아데닌), T(타이민), G(구아닌), C(싸이토신) 4가지가 있지만, 이것을 2가지식 묶게 되면 

AA, AT, AG, AC, TA, TT, TG, TC, GA, GT, GG, GC, CA, CT, CG, CC 로 만들어지고 3가지로 묶으면

AAA, AAT, AAG 등으로 나오게 됩니다

그럼 첫번째로 기본적인 ATGC의 갯수를 출력해봅시다, 여기에서 사용할 함수는 count() 함수를 사용하며

count(변수, 출력할 자릿수) 의 형태로 사용합니다

help('count') 를 입력하게 되면 더 상세한 정보를 알 수 있을것이라고 합니다

In [13]:
count(myseq, 1)


   a    c    g    t 
3426 2240 2770 2299 

In [14]:
count(myseq, 2)


  aa   ac   ag   at   ca   cc   cg   ct   ga   gc   gg   gt   ta   tc   tg   tt 
1108  720  890  708  901  523  261  555  976  500  787  507  440  497  832  529 

In [15]:
count(myseq, 3)


aaa aac aag aat aca acc acg act aga agc agg agt ata atc atg att caa cac cag cat 
388 237 280 203 294 176  94 156 334 184 237 135 141 137 292 138 273 198 227 203 
cca ccc ccg cct cga cgc cgg cgt cta ctc ctg ctt gaa gac gag gat gca gcc gcg gct 
236 106  60 121  70  51  69  71 131 121 186 116 346 197 249 184 169 129  58 144 
gga ggc ggg ggt gta gtc gtg gtt taa tac tag tat tca tcc tcg tct tga tgc tgg tgt 
354 132 160 141  73 106 192 136 101  88 133 118 202 112  49 134 218 133 321 160 
tta ttc ttg ttt 
 95 133 162 139 

In [17]:
help('count') # 많아서 여기에 출력되진 않습니다

**DEN-1 (뎅기열 바이러스)에서 3번째 Element (G) 의 갯수를 출력**

간단합니다 위에서 했던것처럼, dengutable 에다가 count 함수를 사용하는데, 우리는 ACGT 중에서 G 의 갯수를 알아볼 것이기에

값을 1을 집어넣고, 3rd Element (G) 라고 하였으니, dengutable[[3]] 을 입력하면 갯수를 알 수 있으리라 봅니다

( 근데요, 숫자 대신에 g 집어넣어도 되더라구요 )

In [7]:
dengui <- count(myseq, 1)
dengui[[3]]

[1] 2770

In [8]:
dengui[['g']]

[1] 2770

### 흥미진진한 문제풀이의 시간! 

#### 코드는 하단에 있습니다

First. What are the last twenty nucleotides of the Dengue virus genome sequence?

Answer : 결론적으로 뉴클레오타이드의 수를 구한후에 거기에서 [마지막숫자 - 20:마지막숫자] 를 출력해주면 되겠습니다

Second. What is the length in nucleotides of the genome sequence for the bacterium Mycobacterium leprae strain TN (accession NC_002677)?

Answer : 맨 위에 그 기다란 코드가 동작하면 모르겠지만, 동작하지 않으니, NCBI에 들가서 NC_002677 를 검색해서 FASTA 파일을 구해서 집어넣은 뒤에 길이를 구하는 length() 함수를 사용하여 보면 됩니다

Third. How many of each of the four nucleotides A, C, T and G, and any other symbols, are there in the Mycobacterium leprae TN genome sequence?

Answer : 그러면 이것도 맨 위에 그 기다란 코드가 동작하지 않으므로... NCBI 에 들어가서 나균(Mycobacterium leprae TN) 에 대한 FASTA 파일을 가져온 후에, count() 함수를 주고 1번 옵션을 주면 됩니다

Fourth. What is the GC content of the Mycobacterium leprae TN genome sequence, when (i) all non-A/C/T/G nucleotides are included, (ii) non-A/C/T/G nucleotides are discarded?

Answer : GC% 를 구하는 일인데, 아까 말했던 나균(Mycobacterium leprae TN)에서의 GC% 를 구하기 위하여 GC() 함수를 사용하고 거기에 100을 곱해주면 Percentage Of GC 를 구할 수 있습니다

Fifth. How many of each of the four nucleotides A, C, T and G are there in the complement of the Mycobacterium leprae TN genome sequence?

Answer : ~~담은 물안에 있다~~ 답은 문제안에 있는데, Complement 해야 하니까 모든 ACTG 를 압축한 후에 그의 대한 갯수를 구하면 되겠습니다

help(comp) 먼저 검색해보면 되겠습니다, 역시 3번째 문제의 연장선입니다

Sixth. How many occurrences of the DNA words CC, CG and GC occur in the Mycobacterium leprae TN genome sequence?

Answer : 간단하잖아요? [['cc']], [['cg']], [['gc']]의 값을 구하면 되겠네요

Last. How many occurrences of the DNA words CC, CG and GC occur in the (i) first 1000 and (ii) last 1000 nucleotides of the Mycobacterium leprae TN genome sequence?

Answer : 역시나 간단한 문제입니다, CC, CG, GC 문제를 첫번째부터의 1000개와 마지막부터의 1000개부터의 값을 구하는 문제이니, 모든 문제의 해답을 합하면 됩니다.

※ Mycobacterium leprae 이 무었일까 하고 검색해보니, 

나균(癩菌, Hansen’s coccus spirilly)은 온대에서 열대 지역에 걸쳐 발견되는, 그람양성 박테리아의 일종이며, 학명은 미코박테리움 레프라에(Mycobacterium leprae)이며 한센병의 병원균이다

In [16]:
# 1번 문제에 대한 답
length(myseq) # 먼저 뉴클레오타이드의 갯수를 구하고
myseq[10715:10735] # 그 후에 -20 한 수를 넣어서 결과값을 낸다

[1] 10735

[1] "g" "c" "t" "g" "t" "t" "g" "a" "a" "t" "c" "a" "a" "c" "a" "g" "g" "t" "t"
[20] "c" "t"

In [17]:
# 2번 문제의 대한 답 - FASTA DATA의 크기가 5메가 정도 됩니다
NC002677 <- read.fasta(file = "NA1.fasta")[[1]]
length(NC002677)

[1] 3268203

In [19]:
# 3번문제에 대한 답
count(NC002677, 1)


     a      c      g      t 
687041 938713 950202 692247 

In [23]:
# 4번문제에 대한 답
GC(NC002677) * 100

[1] 57.79675

In [25]:
# 5번문제에 대한 답
#help('comp')
result <- comp(NC002677)
table(result)

result
     a      c      g      t 
692247 950202 938713 687041 

In [27]:
CountNC002677 <- count(NC002677, 2)
CountNC002677[['cc']]
CountNC002677[['cg']]
CountNC002677[['gc']]

[1] 236971

[1] 306986

[1] 293261

In [28]:
NC002677F1000 <- count(NC002677[0:1000], 2)
NC002677F1000[['cc']]
NC002677F1000[['cg']]
NC002677F1000[['gc']]
NC002677L1000 <- count(NC002677[3266203:3268203], 2)
NC002677L1000[['cc']]
NC002677L1000[['cg']]
NC002677L1000[['gc']]

[1] 82

[1] 92

[1] 63

[1] 157

[1] 153

[1] 147